### PR TITLE
(DE-667) fix(retrieval): improve logic + fixes

### PR DESCRIPTION
**Pull request by cdelabre-pass**
### Select PR template in preview mode:

 - upgrade to 2.15
 - add ScaNN for metrics validation
 - add full validation over dataset instead of sample
 - update val_loss as MEAN_OVER_BATCHES for loss and val_loss comparison
 - update stop_loss to use val_loss
 - add normalization
 - add adam

### CODE REVIEW

File: jobs/ml_jobs/algo_training/requirements.in [hunk 1]
```diff
@@ -5,9 +5,10 @@ mlflow==2.3.2
 numpy==1.26.4
 pandas==1.5.3
 pandas-gbq>=0.17.1
-tensorflow==2.14.0
+tensorflow[and-cuda]==2.15.1
```
Comment (lmontier-pass) [inline jobs/ml_jobs/algo_training/requirements.in:8]: on veut vraiment passe en 2.15 dans ce ticket. Je dirai que le problème ne vient pas de cette version mais plus d'une batch size ou autre...

File: jobs/ml_jobs/algo_training/requirements.txt [hunk 2]
```diff
@@ -460,36 +464,64 @@ numpy==1.26.4
...
-    #   torch
-nvidia-nccl-cu12==2.20.5
+    #   tensorflow
+nvidia-nccl-cu11==2.14.3
     # via torch
-nvidia-nvjitlink-cu12==12.5.82
+nvidia-nccl-cu12==2.16.5
+    # via tensorflow
+nvidia-nvjitlink-cu12==12.2.140
     # via
```
Comment (lmontier-pass) [inline jobs/ml_jobs/algo_training/requirements.txt:520]: à mon avis il faut regénérer le requirements.txt, il a l'air dans un état chelou...

File: jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py [hunk 3]
```diff
@@ -16,61 +16,53 @@ def __init__(
...
+                    else None,
+                )
+                for name, layer in item_features_config.items()
+            },
             embedding_size=embedding_size,
         )
 
+        item_embeddings = item_dataset.map(self.item_model)
+        index_top_k = tfrs.layers.factorized_top_k.BruteForce()
+        index_top_k.index_from_dataset(item_embeddings)
```
Comment (lmontier-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py:58]: il ne manque pas un index_top_k = ? 

parce que soit dans ton code, cela signifie que index_top_k est in place et dans ce cas là, on s'attend à ce que l'ancien code renvoie none 
```
tfrs.layers.factorized_top_k.BruteForce().index_from_dataset(
                    items_dataset.map(self.item_model)
                )
```,
soit on ne fait rien du return

File: jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py [hunk 4]
```diff
@@ -83,30 +75,33 @@ def compute_loss(self, features, training=False):
...
+            ),
+        }
+        if layer_type not in embedding_layers:
             raise ValueError(
                 f"InvalidLayerType: The features config file contains an invalid layer type `{layer_type}`"
             )
+        return embedding_layers[layer_type]
 
 
 class SingleTowerModel(tf.keras.models.Model):
```
Comment (lmontier-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py:107]: on s'en sert de ce modèle?

File: orchestration/dags/jobs/ml/algo_training_two_towers.py [hunk 5]
```diff
@@ -57,8 +57,8 @@
     "instance_name": f"algo-training-two-towers-{ENV_SHORT_NAME}",
     "instance_type": {
         "dev": "n1-standard-2",
-        "stg": "n1-highmem-8",
-        "prod": "n1-highmem-32",
+        "stg": "n1-standard-8",
+        "prod": "n1-standard-16",
```
Comment (lmontier-pass) [inline orchestration/dags/jobs/ml/algo_training_two_towers.py:61]: :heart:

File: jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py (hunk 3 above)
Comment (cdelabre) [reply jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py]: Dans les deux cas il retourne self !

File: jobs/ml_jobs/algo_training/requirements.txt (hunk 2 above)
Comment (cdelabre) [reply jobs/ml_jobs/algo_training/requirements.txt:520]: Pour quelle raison, il a était compile depuis une VM avec GPU justement...

File: jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py [hunk 6]
```diff
@@ -126,19 +140,21 @@ def __init__(
...
-            feature_embeddings,
-            axis=1,
-        )
+        x = tf.concat(feature_embeddings, axis=1)
 
+        x = self._dropout1(x, training=training)
         x = self._dense1(x)
+        x = self._dropout2(x, training=training)
         x = self._dense2(x)
         return x
```
Comment (lmontier-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py:160]: du coup on ne normalise plus ici ? je trouvais ça pas mal car ça nous permet directement de gérer directement le cas du retrieval, où l'on fait des cosine ou des renormalisations. Du coup, je me dis d'autant le faire directement non?

**Review by lmontier-pass — DISMISSED** [review]
